Add option to use Winforms embedded browser #547

hurcane opened the pull request:
Added a code receiver, EmbeddedBrowserCodeReceiver,  that uses a WinForms-based dialog with an embedded browser control. Added an optional parameter to GoogleWebAuthorizationBroker authorization calls so the client can pass in the requested code receiver. Default code receiver remains the LocalServerCodeReceiver.

hurcane commented:
I signed it!

hurcane commented:
I signed it!

hurcane commented:
I signed the CLA on behalf of Tribute Inc.

chrsmith commented:
Sorry for the delay in getting to this. @peleyal can you take a look?

peleyal commented:
I'm not an expert in Winforms, but before I'm jumping into the review, I'm trying to figure it out if we should create a new assembly for WinForms supports or not.
I don't like the fact that we are adding new dependencies like `System.Drawing` and `System.Windows.Forms`, but on the other hand it doesn't really harm. Chris and Don, what are your thoughts?

Anyway, I'll try to get into it before the weekend.
Thanks!
Eyal

hurcane commented:
I had that same concern when I started work on the task. I started out creating a new project, but I was having a lot of difficulty with deciding on an appropriate name, and I was not sure how that would be organized with respect to NuGet packages.

I considered the following facts in finally deciding to add the dialog to the existing assembly.
* The existing methods imply that the platform is for installed apps, and WinForms naturally fits into the environment of installed apps.
* The project already had two sample receivers.
* WinForms is automatically included with the .NET platform.
* WinForms assemblies won't be loaded unless the embedded browser code receiver is used. Apps that use the prompt code receiver or local server code receiver won't cause the WinForms assemblies to be loaded.
* A WPF control can still be added later without any compatibility issues, with respect to compilation. Users would need to be careful about using the appropriate class for their project, although there are ways to mix and match WPF and Winforms.

Being new to this project, I do not know whether the DotNet4 project is used by Mono-based projects. If it can be used on Mono, then I would have possibly made a different decision, since WinForms is not really a part of the Mono environment.

hurcane commented:
I signed the CLA again. I think there has been some confusion over associated email addresses because I have multiple valid emails with the company.

hurcane commented:
I signed the CLA again. I think there has been some confusion over associated email addresses because I have multiple valid emails with the company.

peleyal commented:
I hear you :+1: 
Good summary. I'm happy that you considered all the above. I agree with you that it's more natural to have it in the .NET 4.0 project and not creating a new one.

Regarding Mono, we don't officially support it, but we don't want to limit ourselves. With what I remember WinForm should be built on Mono, it's WPF that doesn't. right?
+@chrsmith as well, any thoughts?

peleyal commented:
Also @chrsmith, can you take a look in the WinForms specific implementation (since I'm not familiar with WinForms) and I'll review the library's part?

chrsmith commented:
@peleyal Sure thing. I :heart: WinForms.

peleyal commented:
At least someone...  @chrsmith :+1: 

Anyway, @hurcane if you have a sample app that tests it would be great to share it, so Chris would be able to test it himself.

I think that my first comments will be ready by tomorrow.

hurcane commented:
Sorry, but I have some questions that may seem ignorant to somebody experienced with this process. 

In what form would you want a sample app? Since this is part of the authorization process, client secrets need to be provided. I have this integrated into my company's app, but I cannot share that.

Can I just post some sample usage lines, and leave it to Chris to set up a project with necessary references and supply client secrets? Since the change is just to the authorization process, the code to test it won't be too difficult to implement.

peleyal commented:
You should definitely not share your client secret and id. I thought about some sample app, like the ones we have in our samples repository, https://github.com/google/google-api-dotnet-client-samples. I guess you can even have a PR to change https://github.com/google/google-api-dotnet-client-samples/tree/master/Tasks.WinForms.NoteMgr to work with your new API. Later on, after releasing a new version of the library we will even merge this PR to the samples repository.

We just want to test that the new code works. What happens when the users accept the permissions, what happens when they decline, etc.

peleyal commented:
And your question is totally valid. Let me know if my suggestion sounds good to you :)

hurcane commented:
Thanks so much for the links, @peleyal. I had not discovered the samples repo. For sure I can modify the WinForms sample and eventually submit a pull request. For testing purposes, will @chrsmith simply clone my fork of both repos?

In addition to testing the Allow and Deny responses, I had two other use cases:
* Closing the authorization window without clicking Allow or Deny.
* Switching to a different Google account before authorizing.

Closing the window is the use case that led me to make these changes. The local server code receiver has a bug (issue #508) that I could not accept. I initially planned on fixing that bug by implementing a timeout mechanism, but my lack of experience with async/await programs prevented me from being able to figure it out.

In fact, the embedded browser authorization is synchronous from the perspective of the installed application. The calling code will be unresponsive until the browser dialog has closed. Personally, I like the synchronous style as a user, but I would understand if you want the calling app to still be responsive, and eliminate the modal dialog.

I was wondering how I am going to get the sample to use my locally changed version of the library. It looks like I'll need to educate myself on creating a NuGet package and using a local package source for the samples.

I might do some other experimentation with the other samples targeting the net40 framework. I see there is a VB console app for the calendar, and a WPF task list sample. I think it would be interesting to see how the WinForms dialog works with those samples, too.

peleyal commented on Src/Support/GoogleApis.Auth.DotNet4/OAuth2/AuthorizationDialog.cs:
```diff
@@ -0,0 +1,47 @@
+﻿using System;
```
Please add the Google header, e.g. https://github.com/google/google-api-dotnet-client/blob/master/Src/Support/CommonAssemblyInfo.cs
Don't forget to change the year to 2016.

hurcane replied:
Acknowledged and updated.

peleyal commented on Src/Support/GoogleApis.Auth.DotNet4/OAuth2/AuthorizationDialog.cs:
```diff
@@ -0,0 +1,47 @@
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+using Google.Apis.Auth.OAuth2.Responses;
```
Space between System and Google usings.
In general our using statements look like:

using System
[empty line]
using 3rd party libraries, e.g. Json.
[empty line]
using Google....

All sections are sorted alphabetically.

hurcane replied:
Acknowledged and updated. Several of these using statements were unnecessary (automatically added by the Visual Studio Designer). I removed the ones that were not needed.

peleyal commented on Src/Support/GoogleApis.Auth.DotNet4/OAuth2/AuthorizationDialog.cs:
```diff
@@ -0,0 +1,47 @@
+                return responseUrl;
+            }
+        }
+        /// <summary> Constructs a new authentication broker dialog. </summary>
```
No need in space before Constructs and after dialog.

hurcane replied:
I also added a blank line between the property and the constructor method.

peleyal commented on Src/Support/GoogleApis.Auth.DotNet4/OAuth2/AuthorizationDialog.cs:
```diff
@@ -0,0 +1,47 @@
+            // Set the browser to the initial uri.
+            authWebBrowser.Navigate(authUri);
+            // Need to track when the results has been received.
+            authWebBrowser.DocumentCompleted += authWebBrowser_DocumentCompleted;
```
1. Will this line track both accepting and denying the permissions?
2. What happen when the user closes the window?
3. Is there other use case we should think of?

hurcane replied:
1. The DocumentCompleted event fires every time a URL is finished loading. For accept and deny, it will fire with a URL referencing localhost and the query portion of the response is assumed to be Google's AuthorizationCodeResponseUrl.
1. When the form is closed, the caller (EmbeddedBrowserCodeReceiver) has code that checks whether the response URL has been set. If the response has not been set, the task is set to the canceled state.
1. Something I have not tried is checking if the user gets a "localhost" response that is not a proper Google authorization response. This could happen, for example, if the user were to make the browser navigate to a localhost web server. I'm not certain if this is even possible since the embedded browser's navigation controls are disabled. Another potential problem is that a man-in-the-middle attack is sending the user to a malicious localhost. These would not be common scenarios, I hope.

In any case, I would expect that an exception will be thrown by the AuthorizationCodeResponseUrl constructor, which will be caught by the code receiver and associated with the task's result.

hurcane replied:
Another use case is that the user might choose to navigate to a different Google account than the default one that is selected when the browser window loads.

LindaLawton replied:
May I ask why you don't just expose the authentication URI to the user and let anyone who is using windows forms just implement this code themselves?  

I really don't like the idea of adding the windows forms dependency to something not a lot of people are going to be using.

hurcane replied:
@LindaLawton I did some desk checking of what would be required to expose the Authorization URI to the user and allow them to use their own in-process browser to handle receiving the response code from Google.
## Setting context of my thought process
The entire task is a sequence that starts with a code request and ends with a response code token. Everything in-between is currently handled entirely by the library. There are a lot of async operations, but the final result is still procedural since we will always await on the response code.
## Current flow
This is my understanding of the current flow, from a higher level, with many ancillary details omitted.

1. AuthorizationCodeInstalledApp.AuthorizeAsync gets the authorization URI from an IAuthorizationCodeFlow instance.
1. AuthorizationCodeInstalledApp.AuthorizeAsync calls (and awaits) ReceiveCodeAsync on ICodeReceiver.
 * This happens to be an instance of EmbeddedBrowserCodeReceiver.
 * The authorization URI is passed as a parameter
1. EmbeddedBrowserCodeReceiver.ReceiveCodeAsync launches (and awaits, since it is modal) the WinForms dialog
1. WinForms dialog eventually closes (we'll assume it set the response token)
1. EmbeddedBrowserCodeReceiver.ReceiveCodeAsync sets the task result and exits.
1. AuthorizationCodeInstalledApp.AuthorizeAsync exchanges the token for an access token and saves the access token.
 * No further UI is involved for this process.
 * Native HTTP classes are used.

## Options to eliminate WinForms dependency
One option to eliminate the WinForms dependency is replace the call to the dialog with a hook. Since this code is executing a couple of layers internally, one option would be to have an event that the caller handles. The event arg properties would include the authorization URI and a response token. The handler would be expected to set the response token within the event handler.

Another way of doing this might be to make the receiver abstract and require a subclasse, which the user would define, to implement a LaunchBrowser method that returns the authorization response code.

Getting out a bigger hammer, another option would be to move the embedded browser code receiver completely, and put it in the samples repository. In that case, this pull request simply becomes adding the optional parameter to pass in a code receiver to AuthenticationCodeInstalledApp.
## Other thoughts
If we take the dialog form out of the project, but keep the receiver class, I would want to change its name because it would no longer reflect what it does. I would probably change it to BYOBCodeReceiver (i.e. Bring Your Own Browser). 

Is the concern about only Windows Forms, or is it about adding any kind of UI dependency to the project for the "installed app" platform? If I had used a WPF Window instead, would you have the same concern?

I must respectfully challenge the statement that Windows Forms is not something a lot of people are going to be using. While WPF is the shinier technology for new Windows applications, WinForms is still very much in use in the corporate environment. As Richard Campbell of .NET Rocks often says, "WinForms is not dead, it's just done." That is, It's not getting any new features, but it still works.

I will say that I don't think the other current alternatives for Windows-installed apps (LocalServerCodeReceiver and PromptCodeReceiver) are very good options. The prompt receiver really only works in a console app. LocalServerCodeReceiver has a nasty bug that hangs the app if the user closes the external browser (the receiver waits forever for a response that is never coming). The external browser has the problem that it cannot reliably close the browser window, which makes more work for the end user. When I tested with Chrome, I always had to manually close the browser tab after approving the request.

I would also make the case that this is not a burdensome dependency. It is part of the framework, so adding it imposes no new installation requirements. At runtime, the assemblies are only loaded into memory if somebody chooses to use this code receiver. For all other users, there will be no memory changes, other than those due to the new classes in the Platform Services DLL.

LindaLawton replied:
If I had used a WPF Window instead, would you have the same concern?
       Yes, adding any UI dependency to the DLL would have given me the same concern.   IMO the client lib should not be for UI it should just be for the background functionality of accessing the APIs. 

I must respectfully challenge the statement that Windows Forms is not something a lot of people are going to be using.
      I only use windows forms, my wording is probably off.   While I have used WPF 90% of what i do is either console apps or win forms.    

Unfortunately my .Net skills are not IMO up to the challenge of adding a change like this, nor can I comment on your approach.  I will leave that to the .net gurus on the client lib.    However I like the sounds of your BYOB idea.  

BTW:   I made my own class to handle Auth I can then feed the refresh token to the Datastore manually.   [Google APIS auth Simple](http://www.daimto.com/google-apis-auth-simple/)  basically because I needed to be able to launch my own browser from forms.

hurcane replied:
> Yes, adding any UI dependency to the DLL would have given me the same concern. IMO the client lib should not be for UI it should just be for the background functionality of accessing the APIs.

As a counterargument, there is already precedent for including UI elements in the library. The Windows Phone platform includes UI controls to present the browser to the user from within the library. The references to System.Windows.Controls are UI dependencies.

chrsmith replied:
Given the amount of discussion this humble constructor raised, perhaps we should document the code's behavior and/or expectations in a comment.

peleyal commented on Src/Support/GoogleApis.Auth.DotNet4/OAuth2/EmbeddedBrowserCodeReceiver.cs:
```diff
@@ -0,0 +1,75 @@
+namespace Google.Apis.Auth.OAuth2
+{
+    /// <summary>
+    /// OAuth 2.0 verification code receiver for Windows Phone that opens an embedded Google account form to enter the
```
Please update the comment.

hurcane replied:
Acknowledged and updated.

peleyal commented on Src/Support/GoogleApis.Auth.DotNet4/OAuth2/EmbeddedBrowserCodeReceiver.cs:
```diff
@@ -0,0 +1,75 @@
+        /// <param name="url"></param>
+        /// <param name="taskCancellationToken"></param>
+        /// <returns></returns>
+        public Task<AuthorizationCodeResponseUrl> ReceiveCodeAsync(AuthorizationCodeRequestUrl url,
```
Please delete this comment. We use the inherited comments (from ICodeReceiver).

hurcane replied:
Acknowledged and updated.

peleyal commented on Src/Support/GoogleApis.Auth.DotNet4/OAuth2/EmbeddedBrowserCodeReceiver.cs:
```diff
@@ -0,0 +1,75 @@
+                return tcs.Task;
+            }
+            catch (Exception ex)
+            {
```
What exceptions do we expect here? Can we be more specific than catching a general exception?

hurcane replied:
There is no specific exception that I can think of to handle differently. With web browsing, there could be any number of potential exceptions that would occur. This general catch was based on the workflow of the existing AuthorizationCodeBroker, which is also implementing ICodeReceiver. I assumed that since the existing ICodeReceiver implementation was mapping all exceptions to Task.SetException, that this new implementation should do the same thing.

peleyal commented on Src/Support/GoogleApis.Auth.DotNet4/OAuth2/GoogleWebAuthorizationBroker.cs:
```diff
@@ -1,5 +1,5 @@
 ﻿/*
-Copyright 2013 Google Inc
+Copyright 2016 Google Inc
```
No need to update the year. We only write current year when we create a file.

hurcane replied:
Acknowledged and updated.

peleyal commented:
First review; looks good.
I'll wait for you to address the comments and for Chris to actually test it before having the next review.

Regarding your comments:
"I was wondering how I am going to get the sample to use my locally changed version of the library. It looks like I'll need to educate myself on creating a NuGet package and using a local package source for the samples."
>> You don't need to publish\create new NuGet packages. I would just add a local reference to the Google.Apis assemblies you are changing. Then Chris would be able to do the same.
Before releasing a new library, we would be able to actually merge the PR, using the new Google.Apis.Auth version.
I'll leave the details for Chris, but my main point, don't worry about it. Focus about creating the right change and a sample code for it. We will take it from there.

"I might do some other experimentation with the other samples targeting the net40 framework. I see there is a VB console app for the calendar, and a WPF task list sample. I think it would be interesting to see how the WinForms dialog works with those samples, too."
>> I think that you should change only the sample I mentioned because it's the only one that is for WinForms. The others don't run on WinForms and there is NO point in using a WinForm dialog for them.

peleyal commented:
@hurcane, what's the company name you used to sign the CLA? I want to verify it... (an alternative is to sign using you email address that you use in GitHub)

hurcane commented:
The company name is Tribute, Inc. from Hudson, OH USA. I did a couple incorrect submissions using my email as the signer. This past Friday, I submitted the CLA again with the owner's name and email. He signed it, but I don't think we have gotten a confirmation yet.

hurcane commented:
Should I submit a pull request now for the changes in my samples fork? Or can testing be done using a clone of my fork, and the pull request can be submitted after this pull request is accepted?

My fork of the samples repo: https://github.com/hurcane/google-api-dotnet-client-samples/

hurcane commented:
We have received a confirmation of a completed CLA for Tribute Inc.

peleyal commented:
@chrsmith can you please take a look in this one (WinForm wise)?

chrsmith commented on Src/Support/GoogleApis.Auth.DotNet4/OAuth2/AuthorizationDialog.cs:
```diff
@@ -0,0 +1,58 @@
+    {
+        private AuthorizationCodeResponseUrl responseUrl;
+
+        public AuthorizationCodeResponseUrl ResponseUrl
```
You can use a C# auto property here:

public AuthorizationCodeResponseUrl { get; protected set; }

chrsmith commented:
The WinForms code looks reasonable to me. (I didn't build and see what the .resx file looks like in its full GDI+ enabled glory.) If @peleyal is happy with the rest of the code, I'm cool with giving it a :+1:

peleyal commented:
1. @chrsmith Can you run a WinForm sample using the new code? I prefer us to test it before submitting it.

2. Do we all agree that we should add it to the current Auth assembly and not to a new NuGet package for WinForm extensions? It's a tough call, and setting up a new project + NuGet configuration might be an overkill here (similar approach was taken for ASP.NET MVC, https://www.nuget.org/packages/Google.Apis.Auth.Mvc).
IMO setting up a project just for WinForm is better, because WinForm isn't used by 99% of Google.Apis.Auth users and it's a better separation.
Sorry that I'm bringing it up again, but I want us all to be on the same page regarding this one.

Thanks!
Eyal

hurcane commented:
Our internal testing has revealed a bug. The project works on machines that have IIS installed, even if it is not active. It is failing on machines that don't have IIS. I will be addressing that, and I could move the WinForms-dependent classes into a separate DLL, following the MVC project model.

hurcane commented:
@peleyal @chrsmith  @LindaLawton  I have moved the WinForms dependencies into a separate project. In addition, I have a fork of the google-api-dotnet-client-samples project that includes changes to the WinForms Notes sample to allow testing the embedded browser in the WinForms project.

In addition, I corrected the bug in the previous commits. The redirect URL was going to localhost by default, but WinForms projects should use the "out-of-band" redirect URL. The project has been modified to work out-of-band.

Designing NuGet specifications is new to me. I hope that I did it correctly for the new project in the solution.

hurcane commented:
@peleyal @chrsmith @LindaLawton Is anybody willing to view the updates I have made to this pull request?

chrsmith commented:
@hurcane I am sorry for this taking so long. I swear to <emoji for diety missing> will make the time to review this code tomorrow.

Moving the WinForms bits to a separate project removes most of the concerns I would have had -- chiefly bringing in dependencies that might cause problems on specific platforms. I'll take a final look and let you know in less than 20 hours. (Again, sorry.)

chrsmith commented on Src/Support/GoogleApis.Auth.WinForms/OAuth2/WinForms/AuthorizationDialog.cs:
```diff
@@ -0,0 +1,63 @@
+                this.Close();
+            }
+            else
+                System.Diagnostics.Trace.TraceInformation(browser.Url.AbsolutePath);
```
Please use curly braces to denote a block after the else-expression, even if they aren't strictly needed in this case. e.g.

````
} else {
    System.Diagnostics.Trace.TraceInformation(browser.Url.AbsolutePath);
}
````

hurcane replied:
Actually, I will remove this else because I only had it there to diagnose that I was handling all of the appropriate values.

From: Chris Smith [mailto:notifications@github.com]
Sent: Friday, March 18, 2016 1:46 PM
To: google/google-api-dotnet-client
Cc: Don Benson
Subject: Re: [google-api-dotnet-client] Add option to use Winforms embedded browser #547 (#665)


In Src/Support/GoogleApis.Auth.WinForms/OAuth2/WinForms/AuthorizationDialog.cs<https://github.com/google/google-api-dotnet-client/pull/665#discussion_r56695329>:

> +        }

> +

> +        private void authWebBrowser_TitleChanged(object sender, EventArgs e)

> +        {

> +            // The result always comes back

> +            var browser = (WebBrowser)sender;

> +            if (browser.Url.AbsolutePath == new Uri( GoogleAuthConsts.ApprovalUrl).AbsolutePath)

> +            {

> +                // Per API documentation, the reponse is after the first space in the title.

> +                var query = browser.DocumentTitle.Substring(browser.DocumentTitle.IndexOf(" ") + 1);

> +                ResponseUrl = new AuthorizationCodeResponseUrl(query);

> +                // Dialog is no longer neeed.

> +                this.Close();

> +            }

> +            else

> +                System.Diagnostics.Trace.TraceInformation(browser.Url.AbsolutePath);

Please use curly braces to denote a block after the else-expression, even if they aren't strictly needed in this case. e.g.

} else {

    System.Diagnostics.Trace.TraceInformation(browser.Url.AbsolutePath);

}

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/google/google-api-dotnet-client/pull/665/files/c62970461d4360d8749299c81daefc647f3de229#r56695329>

________________________________
CONFIDENTIALITY NOTICE: This email and any files transmitted with it are the property of Tribute Inc. and/or its affiliates. The contents of this communication are confidential and may contain information that is privileged and/or exempt from disclosure under applicable law. It is intended solely for use of the individual or entity to whom this email is addressed. If you are not one of the named recipient(s) or otherwise have reason to believe that you have received this message in error, please notify the sender and immediately delete this message and any attachments. Any unauthorized use, retention, dissemination, forwarding, printing, or copying of this email is strictly prohibited.

chrsmith commented on Src/Support/GoogleApis.Auth.WinForms/OAuth2/WinForms/AuthorizationDialog.cs:
```diff
@@ -0,0 +1,63 @@
+            var browser = (WebBrowser)sender;
+            if (browser.Url.AbsolutePath == new Uri( GoogleAuthConsts.ApprovalUrl).AbsolutePath)
+            {
+                // Per API documentation, the reponse is after the first space in the title.
```
Per what API documentation? Can you provide a link?

hurcane replied:
The API example I found was for YouTube. I think that should still apply since the authorization servers are independent, right?

https://developers.google.com/youtube/2.0/developers_guide_protocol_oauth2#OAuth2_Installed_Applications_Flow

In the Installed Application section, step 4 contains the details that show how the response is returned.  Do you want this link to go into the code?

From: Chris Smith [mailto:notifications@github.com]
Sent: Friday, March 18, 2016 1:47 PM
To: google/google-api-dotnet-client
Cc: Don Benson
Subject: Re: [google-api-dotnet-client] Add option to use Winforms embedded browser #547 (#665)


In Src/Support/GoogleApis.Auth.WinForms/OAuth2/WinForms/AuthorizationDialog.cs<https://github.com/google/google-api-dotnet-client/pull/665#discussion_r56695377>:

> +            // Set the browser to the initial uri.

> +            authWebBrowser.Navigate(authUri);

> +            // Need to track when the results has been received.

> +            // Installed applications use an out-of-band redirect URI.

> +            // When the user clicks Accept or Deny, the redirected response will

> +            // be in the title of the web page.

> +            authWebBrowser.DocumentTitleChanged += authWebBrowser_TitleChanged;

> +        }

> +

> +        private void authWebBrowser_TitleChanged(object sender, EventArgs e)

> +        {

> +            // The result always comes back

> +            var browser = (WebBrowser)sender;

> +            if (browser.Url.AbsolutePath == new Uri( GoogleAuthConsts.ApprovalUrl).AbsolutePath)

> +            {

> +                // Per API documentation, the reponse is after the first space in the title.

Per what API documentation? Can you provide a link?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/google/google-api-dotnet-client/pull/665/files/c62970461d4360d8749299c81daefc647f3de229#r56695377>

________________________________
CONFIDENTIALITY NOTICE: This email and any files transmitted with it are the property of Tribute Inc. and/or its affiliates. The contents of this communication are confidential and may contain information that is privileged and/or exempt from disclosure under applicable law. It is intended solely for use of the individual or entity to whom this email is addressed. If you are not one of the named recipient(s) or otherwise have reason to believe that you have received this message in error, please notify the sender and immediately delete this message and any attachments. Any unauthorized use, retention, dissemination, forwarding, printing, or copying of this email is strictly prohibited.

chrsmith commented:
Hey, sorry for taking so long to get to this PR. We (Google) are trying to improve the overall level of support for our client libraries, and so we are still working out things like release process, PR process, etc.

I spoke with some people in the office, and thought more about what we want to do in terms of support. I do not think we should merge this pull request into the repo.

While this PR may improve the client-side app experience when using Google APIs, it also complicates the test/release matrix. e.g. how well is WinForms supported on other platforms if at all? How can we best test this library for future releases? Even if the code is in another csproj, it still adds complexity we need to support in the future.

However, I'm happy to take the refactorings to enable your change. (e.g. exposing an `ICodeReceiver` parameter in `GoogleApis.Auth.DotNet4`.) But beyond that I'd encourage you to create your own GitHub repo for `GoogleApis.Auth.WinForms` portion. I'd be happy to link to it in the documentation for this project.

I will close this PR out. Please open another PR specifically for the refactoring to enable custom `ICodeRecivers`. I am sorry it took so long to bring this PR to resolution.
